plugin/evm/validators/state.go

```diff
@@ -0,0 +1,313 @@
+				return err
+			}
+			if err := batch.Put(vID[:], dataBytes); err != nil {
+				return err
+			}
+		} else { // deleted
+			if err := batch.Delete(vID[:]); err != nil {
+				return err
+			}
+		}
+		// we're done, remove the updated marker
+		delete(s.updatedData, vID)
+	}
+	return batch.Write()
+}
+
+// SetStatus sets the active status of the validator with the given vID
+func (s *state) SetStatus(vID ids.ID, isActive bool) error {
+	data, exists := s.data[vID]
+	if !exists {
+		return database.ErrNotFound
+	}
+	data.IsActive = isActive
+	s.updatedData[vID] = true
+
+	for _, listener := range s.listeners {
+		listener.OnValidatorStatusUpdated(vID, data.NodeID, isActive)
+	}
+	return nil
+}
+
+// GetStatus returns the active status of the validator with the given vID
+func (s *state) GetStatus(vID ids.ID) (bool, error) {
+	data, exists := s.data[vID]
+	if !exists {
+		return false, database.ErrNotFound
+	}
+	return data.IsActive, nil
+}
+
+// GetValidationIDs returns the validation IDs in the state
+func (s *state) GetValidationIDs() set.Set[ids.ID] {
+	ids := set.NewSet[ids.ID](len(s.data))
+	for vID := range s.data {
+		ids.Add(vID)
+	}
+	return ids
+}
+
+// GetValidatorIDs returns the validator IDs in the state
+func (s *state) GetValidatorIDs() set.Set[ids.NodeID] {
+	ids := set.NewSet[ids.NodeID](len(s.index))
+	for nodeID := range s.index {
+		ids.Add(nodeID)
+	}
+	return ids
+}
+
+// RegisterListener registers a listener to the state
+// the listener will be notified of current validators via OnValidatorAdded
+func (s *state) RegisterListener(listener StateCallbackListener) {
+	s.listeners = append(s.listeners, listener)
+
+	// notify the listener of the current state
+	for vID, data := range s.data {
+		listener.OnValidatorAdded(vID, data.NodeID, uint64(data.startTime.Unix()), data.IsActive)
+	}
+}
+
+// parseValidatorData parses the data from the bytes into given validatorData
+func parseValidatorData(bytes []byte, data *validatorData) error {
+	if len(bytes) != 0 {
+		if _, err := vdrCodec.Unmarshal(bytes, data); err != nil {
+			return err
+		}
+	}
+	data.lastUpdated = time.Unix(int64(data.LastUpdated), 0)
+	data.startTime = time.Unix(int64(data.StartTime), 0)
+	return nil
+}
+
+// Load the state from the disk
+func (s *state) loadFromDisk() error {
+	it := s.db.NewIterator()
+	defer it.Release()
+	for it.Next() {
+		vIDBytes := it.Key()
+		vID, err := ids.ToID(vIDBytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse validator ID: %w", err)
+		}
+		vdr := &validatorData{
+			validationID: vID,
+		}
+		if err := parseValidatorData(it.Value(), vdr); err != nil {
+			return fmt.Errorf("failed to parse validator data: %w", err)
+		}
+		if err := s.putData(vID, vdr); err != nil {
+			return err
+		}
+	}
+	return it.Error()
+}
+
+func (s *state) putData(vID ids.ID, data *validatorData) error {
+	if _, exists := s.data[vID]; exists {
+		return fmt.Errorf("validator data already exists for %s", vID)
+	}
+	// should never happen
```

Comment:
is there any difference between these 2 cases in that they are always unexpected? (should the comment apply to one/both?)
Since it seems this error could be returned from the exported function `AddNewValidator`

Reply:
the first check (unique `vID`) should always return before second `nodeID` check. the first check is expected in case where `loadFromDisk` is called twice or where the `AddNewValidator` is called multiple times with the same parameters. 

It's actually possible to have the same `nodeID` under a different `vID` but not at the same time. so they're not "always" mutually unique to each other. removed that comment.